### ISSUE TITLE
Develop a swarm-plugin for Qiime 2

### ISSUE THREAD

**Issue by frederic-mahe** [torognes/swarm#89]
[Qiime 2](http://2.qiime.org/) now offers an interface for third-party plugins. The [plugin creation](http://2.qiime.org/docs/Creating-a-QIIME-2-plugin.html) does not seem complicated: the plugin is a python 3 wrapper presenting some or all the functionalities of swarm.

**Comment by torognes** [comment]
Sounds like a good idea!

**Comment by stheil15** [comment]
Any update ?

**Comment by colinbrislawn** [comment]
With swarm 3.0 fast approaching (#122), the increasing popularity of Exact Sequence Variants, and the publication of the Qiime 2 paper, this might be the _perfect_ time to build a q2-swarm plugin.

🚀
Colin

**Comment by frederic-mahe** [comment]
@colinbrislawn you are right, but I don't really know where to begin. Would you help me kickstart that plugin?

**Comment by colinbrislawn** [comment]
Thanks @frederic-mahe! I'm honored you reached out to me, but I'm not sure where to begin either. I guess I would look to the q2-vsearch plugin as a template, then build from there.  https://github.com/qiime2/q2-vsearch 

@thermokarst, could you make us an official q2-swarm repo and invite us as contributors?

**Comment by thermokarst** [comment]
Hey there @colinbrislawn! This plugin idea sounds really interesting, and good news, no need for us to make you a repo! Since QIIME 2 is decentralized, you can create the plugin wherever you want, then you can share it with users by registering it at the [QIIME 2 Library](https://library.qiime2.org)! The Library entry can contain instructions letting users know how to get your plugin and install it.

**Comment by colinbrislawn** [comment]
Summary of steps in [Fred's-metabarcoding-pipeline](https://github.com/frederic-mahe/swarm/wiki/Fred's-metabarcoding-pipeline), as I understand it, and what's already wrapped in Qiime2:

| program                  | idea               | existing q2 plugin | what                                                             |
|--------------------------|--------------------|--------------------|------------------------------------------------------------------|
|                          |                    | qiime tools import |                                                                  |
| cutadapt                 |                    | q2-cutadapt        |                                                                  |
| vsearch                  | fastq_mergepairs   | q2-vsearch         |                                                                  |
| vsearch                  | fastq_filter       | extend q2-vsearch   | add --fastq_filter   |
| vsearch                  | per-sample derep   | extend q2-vsearch   |                                                                  |
| sed                      | per-read quality   | none               | the lowest expected error rate observed for each unique sequence |
| vsearch                  | global derep       | q2-vsearch         |                                                                  |
| swarm                    | make ASVs!         | none               | then sort with vsearch                                           |
| vsearch                  | uchime_denovo      | q2-vsearch         |                                                                  |
| OTU_contingency_table.py | make feature table |                    |                                                                  |

This is a fully featured pipeline that differs from what's already in Qiime2 in a number of ways. Specifically the per-sample derep...

One easy way forward is to make a q2-swarm plugin that replaces only the [vsearch cluster-features-de-novo](https://docs.qiime2.org/2024.2/plugins/available/vsearch/cluster-features-de-novo/).

This is in contrast to the DADA2 plugin that implements its full, unique SOP.
This may be more powerful; per-sample derep and per-feature quality are interesting ideas!

Either way, adding ` --fastq_filter` to q2-vsearch seems like a natural first step.

**Comment by frederic-mahe** [comment]
I should have pointed that sooner, here is [my current swarm-based pipeline](https://github.com/frederic-mahe/fred-metabarcoding-pipeline/blob/master/01_merge_trim_and_dereplicate/merge_trim_and_dereplicate.org).

The way the pipeline is described (and scripts numbered) might be confusing. The beginning is quite similar to the old pipeline you were referring to: `--fastq_filter` is indeed required.

> One easy way forward is to make a q2-swarm plugin that replaces only the [vsearch cluster-features-de-novo](https://docs.qiime2.org/2024.2/plugins/available/vsearch/cluster-features-de-novo/).

I realize that replicating the whole pipeline in Qiime2 might not be easy, so I agree we should aim for an easier first target.

`swarm` has three major modes: `--differences 0` (dereplication), `--differences 1` (fast, high-resolution clustering),  `--differences 2+` (slower, lower-resolution clustering)

In my own work, I only use `--differences 1`, with the `--fastidious` option when clustering the whole project (`clustering()`), or without the `--fastidious` option when working at the sample level (`list_local_clusters()`).

```sh
list_local_clusters() {
    # retain only clusters with more than 2 reads
    # (do not use the fastidious option here)
    ${SWARM} \
        --differences 1 \
        --threads "${THREADS}" \
        --usearch-abundance \
        --log /dev/null \
        --output-file /dev/null \
        --statistics-file - \
        "${SAMPLE}.fas" | \
        awk 'BEGIN {FS = OFS = "\t"} $2 > 2' > "${SAMPLE}.stats"
}
```
```sh
clustering() {
    # swarm 3 or more recent
    "${SWARM}" \
        --differences 1 \
        --fastidious \
        --usearch-abundance \
        --threads "${THREADS}" \
        --internal-structure "${OUTPUT_STRUCT}" \
        --output-file "${OUTPUT_SWARMS}" \
        --statistics-file "${OUTPUT_STATS}" \
        --seeds "${OUTPUT_REPRESENTATIVES}" \
        "${FINAL_FASTA}" 2> "${OUTPUT_LOG}"
}
```
The input file is a dereplicated fasta file with abundance annotations (`;size=123[;]`) (option `--usearch-abundance`), and only ACGT nucleotides. The command line options listed in these shell functions are the most relevant, at least in my opinion.

**Comment by colinbrislawn** [comment]
Thank you, this is extremely helpful!

I like the idea of starting small with the q2-swarm plugin.

>I only use --differences 1, with the --fastidious option when clustering the whole project (`clustering()`)...

Naturally!

I'm not sure how best to track feature counts through per-sample derep and clustering.

I understand what per-sample derep does and why it's faster to do this double-derep step.
For some reason, I can't wrap my head around that q2-type it should be. Is this just another feature table?
(I also had trouble with this [last time](https://github.com/torognes/vsearch/issues/129#issuecomment-146943277).)

We could get counts for the feature table by remapping reads, like [we did historically](https://drive5.com/usearch/manual6/otupipe.html), but that loses the efficiency of the pre-sample derep and ignores the internal structure of the swarms.
